typeclass-experiments/src/Gen/Rust.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Gen.Rust where
+
+import Data.Set (Set)
+import Data.Set qualified as S
+import Data.Text (Text)
+import Data.Text qualified as T
+
+import Common.Types
+import Gen.Generator
+
+import Debug.Trace
+
+-- these are just convenience pattern variables
+k, v, _a, _l, _x, _xs, _name, _vars :: Pat
+k = VarP "k"
+v = VarP "v"
+_a = VarP "a"
+_l = VarP "l"
+_x = VarP "x"
+_xs = VarP "xs"
+_name = VarP "name"
+_vars = VarP "vars"
+_nil = VarP "nil"
+
+{- utilities for formatting code -}
+brackets :: Text -> Text
+brackets t = "{\n" <> t <> " }"
+
+indent :: Text -> Text
+indent t = "  " <> t
+
+indent2 :: Text -> Text
+indent2 = indent . indent
+
+parens :: Text -> Text
+parens t = "(" <> t <> ")"
+
+break :: Text -> Text
+break t = t <> "\n"
+
+-- error throwing utility
+matchFail :: Pat -> Pat -> Either GenError a
+matchFail p1 p2 = Left $ MatchFail p1 p2
+
+{- Primitives -}
+rustInt :: Gen
+rustInt = primGen IntP "i64"
+
+rustString :: Gen
+rustString = primGen StringP "String"
+
+rustBool :: Gen
+rustBool = primGen BoolP "bool"
+
+rustRef :: Gen
+rustRef = Gen (RefP _x) $ \_ pat -> case pat of
+  (RefP (Name n)) -> pure n
+  other -> matchFail (RefP _x) other
+
+{- Functors -}
+rustMaybe :: Gen
+rustMaybe = Gen (Maybe _x) $ \scope pat -> case pat of
+  Maybe t -> do
+    t' <- runGenerator scope t
+    pure $ "Option<" <> t' <> ">"
+  other -> matchFail (Maybe _x) other
+
+rustList :: Gen
+rustList = Gen (List _x) $ \scope pat -> case pat of
+  List t -> do
+    t' <- runGenerator scope t
+    pure $ "Vec<" <> t' <> ">"
```

Comment:
With Rust DSL this would be `pure $ RustBuiltinVector t'`

Reply:
What do we get out of reifying that into a data type? I feel like that just adds a whole lot of code for no real benefit. See above comment about having `Gen` return `(Doc,[Import])`. I think that approach would end up saving us a lot of typing and ultimately be just as maintainable. It just seems like another layer of indirection to me.

Reply:
Clarity, testing, robustness, separation of concern. I think it's useful as looking at the Rust DSL module it would 'teach me' all I need to know what Rust is to a necessary level of detail.

I'm also thinking how easy it would be to make property or even model testing of Lb -> Rust mappings, and also we don't have to rely on tests based on Text matching as much.

That's my argument for this indirection, and generally I'm wary of formatting text anywhere close to 'core' logic, and it's generally recommended to delegate 'text rendering' to a very small and scoped module.